test/storage/projected_row_test.cpp

```diff
@@ -0,0 +1,110 @@
+    storage::ProjectedRow *update =
+        storage::ProjectedRow::InitializeProjectedRow(update_buffer, update_col_ids, layout);
+    StorageTestUtil::PopulateRandomRow(update, layout, null_ratio_(generator_), &generator_);
+    loose_pointers_.push_back(update_buffer);
+
+
+    // generator a binary vector and set nulls according to binary vector. For null attributes, we set value to be 0.
+    std::bernoulli_distribution coin(null_ratio_(generator_));
+    std::vector<bool> null_cols(update->NumColumns());
+    for (uint16_t  i = 0; i < update->NumColumns(); i++) {
+      null_cols[i] = coin(generator_);
+      if (null_cols[i]) {
+        byte * val_ptr = update->AccessForceNotNull(i);
+        storage::StorageUtil::WriteBytes(layout.attr_sizes_[i+1], 0, val_ptr);
+        update->SetNull(i);
+    } else {
+        update->SetNotNull(i);
+      }
+    }
+    // check correctness
+    for (uint16_t i = 0; i < update->NumColumns(); i++) {
+      byte * addr = update->AccessWithNullCheck(i);
+      if (null_cols[i]) {
+        EXPECT_EQ(addr, nullptr);
+        byte * val_ptr = update->AccessForceNotNull(i);
+        EXPECT_EQ(0, storage::StorageUtil::ReadBytes(layout.attr_sizes_[i+1], val_ptr));
+      } else {
+        EXPECT_FALSE(addr == nullptr);
+      }
+    }
+  }
+}
+
+// This test generates randomized block layouts, and checks its layout to ensure
+// that the header, the column bitmaps, and the columns don't overlap, and don't
+// go out of page boundary. (In other words, memory safe.)
+// NOLINTNEXTLINE
+TEST_F(ProjectedRowTests, MemorySafety){
+  const uint32_t num_iterations = 500;
+  for (uint32_t iteration = 0; iteration < num_iterations; iteration++) {
+    // get a random table layout
+    storage::BlockLayout layout = StorageTestUtil::RandomLayout(MAX_COL, &generator_);
+
+    // generate a random projectedRow
+    std::vector<uint16_t> update_col_ids = StorageTestUtil::ProjectionListAllColumns(layout);
+    auto *update_buffer = new byte[storage::ProjectedRow::Size(layout, update_col_ids)];
+    storage::ProjectedRow *update =
+        storage::ProjectedRow::InitializeProjectedRow(update_buffer, update_col_ids, layout);
+    loose_pointers_.push_back(update_buffer);
+
+    EXPECT_EQ(layout.num_cols_ -1, update->NumColumns());
+    void *lower_bound = update->ColumnIds();
+    void *upper_bound = update + update->Size();
+    // check the first value is in bound
+    StorageTestUtil::CheckInBounds(update->AccessForceNotNull(0), lower_bound, upper_bound);
+    // check the rest values memory addresses don't overlapping previous addresses.
+    for (uint16_t col = 1; col < update->NumColumns(); col++) {
+      lower_bound = update->AccessForceNotNull(static_cast<uint16_t >(col - 1));
+      upper_bound = StorageTestUtil::IncrementByBytes(update->AccessForceNotNull(static_cast<uint16_t >(col - 1)),
+                                                      layout.attr_sizes_[col]);
+      // check if the value address is in bound
```

Comment:
not in bound?

The use of CheckNotInBounds is probably correct judging by your comment outside the loop, but maybe it's most straightforward to instead advance the lower bound incrementally and use CheckInBounds against a constant upper bound.

Reply:
I saw that was what you did for TAS test. My concern was that if you fix the upper bound to be the very end, then the bound is too big and the address can still be incorrect even if it's in bound.  
On second thought, I think your way of checking bound is still correct because of some sort of "induction" reason. 

Correct me if I'm wrong.

Reply:
What I'm doing here is to,  for each column value,  identify the memory space used by the previous value, which is [lower_bound, upper_bound) and make sure that the current value's memory address don't start within this range.

Reply:
Never mind. I agree that advancing the lower bound and keeping upper bound constant serve the same purpose and are more straightforward.